# CASE STUDY #1 - DANNY'S DINER 

[![View Main Folder](https://img.shields.io/badge/View-Main_Folder-971901?)](https://github.com/phamdinhtuan200382/8-Week-SQL-Challenge)
[![View Repositories](https://img.shields.io/badge/View-My_Repositories-blue?logo=GitHub)](https://github.com/phamdinhtuan200382/8-Week-SQL-Challenge/tree/main/Case_Study_1_Dannys_Dinner.ipynb-1) 
[![View My Profile](https://img.shields.io/badge/View-My_Profile-green?logo=GitHub)](https://github.com/phamdinhtuan200382)

## Problem Statement
---
Danny seriously loves Japanese food so in the beginning of 2021, he decides to embark upon a risky venture and opens up a cute little restaurant that sells his 3 favourite foods: sushi, curry and ramen. 
Danny wants to use the data to answer a few simple questions about his customers.  He plans on using these insights to help him decide whether he should expand the existing customer loyalty program.


Danny has provided you with a sample of his overall customer data due to privacy issues - but he hopes that these examples are enough for you to write fully functioning SQL queries to help him answer his questions!

## Import SQLite and Load to Use
---

In [5]:
import sqlite3
import datetime

In [6]:
%load_ext sql
%sql sqlite://


The sql extension is already loaded. To reload it, use:
  %reload_ext sql


## Dataset
---
Danny has shared with you 3 key datasets for this case study:
* sales
* menu
* members

This is entity relationship diagram
<p align="center">
<img src="Image\entity_relationship_diagram.png">

## Table 1: sales

The sales table captures all ```customer_id``` level purchases with an corresponding ```order_date``` and ```product_id``` information for when and what menu items were ordered

In [7]:
%%sql
CREATE TABLE sales (
  "customer_id" VARCHAR(1),
  "order_date" DATE,
  "product_id" INTEGER
);

INSERT INTO sales
  ("customer_id", "order_date", "product_id")
VALUES
  ('A', '2021-01-01', '1'),
  ('A', '2021-01-01', '2'),
  ('A', '2021-01-07', '2'),
  ('A', '2021-01-10', '3'),
  ('A', '2021-01-11', '3'),
  ('A', '2021-01-11', '3'),
  ('B', '2021-01-01', '2'),
  ('B', '2021-01-02', '2'),
  ('B', '2021-01-04', '1'),
  ('B', '2021-01-11', '1'),
  ('B', '2021-01-16', '3'),
  ('B', '2021-02-01', '3'),
  ('C', '2021-01-01', '3'),
  ('C', '2021-01-01', '3'),
  ('C', '2021-01-07', '3');

 * sqlite://
Done.
15 rows affected.


[]

In [8]:
%%sql
SELECT * FROM sales

 * sqlite://
Done.


[('A', '2021-01-01', 1),
 ('A', '2021-01-01', 2),
 ('A', '2021-01-07', 2),
 ('A', '2021-01-10', 3),
 ('A', '2021-01-11', 3),
 ('A', '2021-01-11', 3),
 ('B', '2021-01-01', 2),
 ('B', '2021-01-02', 2),
 ('B', '2021-01-04', 1),
 ('B', '2021-01-11', 1),
 ('B', '2021-01-16', 3),
 ('B', '2021-02-01', 3),
 ('C', '2021-01-01', 3),
 ('C', '2021-01-01', 3),
 ('C', '2021-01-07', 3)]

## Table 2: menu
The menu table maps the product_id to the actual product_name and price of each menu item.

In [9]:
%%sql
CREATE TABLE menu (
  "product_id" INTEGER,
  "product_name" VARCHAR(5),
  "price" INTEGER
);

INSERT INTO menu
  ("product_id", "product_name", "price")
VALUES
  ('1', 'sushi', '10'),
  ('2', 'curry', '15'),
  ('3', 'ramen', '12');

 * sqlite://
Done.
3 rows affected.


[]

In [10]:
%%sql
SELECT * FROM menu 

 * sqlite://
Done.


[(1, 'sushi', 10), (2, 'curry', 15), (3, 'ramen', 12)]

## Table 3: members
The final members table captures the join_date when a customer_id joined the beta version of the Danny’s Diner loyalty program.

In [11]:
%%sql
CREATE TABLE members (
  "customer_id" VARCHAR(1),
  "join_date" DATE
);

INSERT INTO members
  ("customer_id", "join_date")
VALUES
  ('A', '2021-01-07'),
  ('B', '2021-01-09');

 * sqlite://
Done.
2 rows affected.


[]

In [12]:
%%sql
SELECT * FROM members

 * sqlite://
Done.


[('A', '2021-01-07'), ('B', '2021-01-09')]

## Case Study Questions
---
Each of the following case study questions can be answered using a single SQL statement:

1. What is the total amount each customer spent at the restaurant?
2. How many days has each customer visited the restaurant?
3. What was the first item from the menu purchased by each customer?
4. What is the most purchased item on the menu and how many times was it purchased by all customers?
5. Which item was the most popular for each customer?
6. Which item was purchased first by the customer after they became a member?
7. Which item was purchased just before the customer became a member?
8. What is the total items and amount spent for each member before they became a member?
9. If each $1 spent equates to 10 points and sushi has a 2x points multiplier - how many points would each customer have?
10. In the first week after a customer joins the program (including their join date) they earn 2x points on all items, not just sushi - how many points do customer A and B have at the end of January?

## Solution
---

### 1. What is the total amount each customer spent at the restaurant?

In [13]:
%%sql
SELECT 
	sales.customer_id,
  SUM(menu.price) AS total_spent
FROM sales
JOIN menu
	ON sales.product_id = menu.product_id
GROUP BY customer_id
ORDER BY customer_id

 * sqlite://
Done.


[('A', 76), ('B', 74), ('C', 36)]

**Answer-01**:

- Customer A spent $76
- Customer B spent $74
- Customer C spent $36

### 2. How many days has each customer visited the restaurant?

In [14]:
%%sql
SELECT customer_id, COUNT(DISTINCT(order_date)) AS no_of_visit
FROM sales
GROUP BY customer_id

 * sqlite://
Done.


[('A', 4), ('B', 6), ('C', 2)]

**Answer-02**:

- Customer A has visited 4 days
- Customer B has visited 6 days
- Customer C has visited 2 days

### 3. What was the first item from the menu purchased by each customer?

In [15]:
%%sql
SELECT s.order_date, s.customer_id, m.product_name
FROM sales s
JOIN menu m
ON s.product_id =m.product_id
GROUP BY s.customer_id
ORDER BY s.customer_id ASC


 * sqlite://
Done.


[('2021-01-01', 'A', 'sushi'),
 ('2021-01-01', 'B', 'curry'),
 ('2021-01-01', 'C', 'ramen')]

**Answer-03**:

First item purchased by:
- Customer A is sushi
- Customer B is curry
- Customer C is ramen

### 4. What is the most purchased item on the menu and how many times was it purchased by all customers?

In [16]:
%%sql
SELECT m.product_name, COUNT(s.product_id) AS  times
FROM sales s
JOIN menu m
ON s.product_id = m.product_id
GROUP BY m.product_name
ORDER BY times DESC
LIMIT 1

 * sqlite://
Done.


[('ramen', 8)]

**Answer-04**:

- Most purchased item was ramen and it was ordered 8 times

### 5.Which item was the most popular for each customer?

In [17]:
%%sql
WITH rank_items AS
(
    SELECT s.customer_id, m.product_name, COUNT(s.product_id),
	RANK() OVER(PARTITION BY s.customer_id ORDER BY COUNT(s.product_id) DESC) AS rank
	FROM sales s
	JOIN menu m
    ON s.product_id = m.product_id
	GROUP BY s.customer_id, m.product_name
)
SELECT r.customer_id, r.product_name as popular_items
FROM rank_items r
WHERE rank = 1

 * sqlite://
Done.


[('A', 'ramen'),
 ('B', 'sushi'),
 ('B', 'ramen'),
 ('B', 'curry'),
 ('C', 'ramen')]

**Answer-05**:

Most popular item for:
- Customer A was Ramen
- Customer B was a tie between all three menu items
- Customer C was ramen

### 6. Which item was purchased first by the customer after they became a member?

In [18]:
%%sql
WITH rank_items AS
(
    SELECT s.customer_id, s.order_date, mem.join_date ,m.product_name,
    RANK() OVER(PARTITION BY s.customer_id ORDER BY s.order_date ASC) AS rank
    FROM sales s
    JOIN menu m
    ON s.product_id = m.product_id
    JOIN members mem
    ON s.customer_id = mem.customer_id
    WHERE s.order_date > mem.join_date
)
SELECT r.customer_id, r.product_name
FROM rank_items r
WHERE r.rank =1

 * sqlite://
Done.


[('A', 'ramen'), ('B', 'sushi')]

**Answer-06**:

Only A and B are in membership and first product ordered after becoming a member for:
- Customer A is ramen
- Customer B is sushi

### 7.Which item was purchased just before the customer became a member?

In [19]:
%%sql
WITH rank_items AS
(
    SELECT s.customer_id, s.order_date, mem.join_date ,m.product_name,
    RANK() OVER(PARTITION BY s.customer_id ORDER BY s.order_date ASC) AS rank
    FROM sales s
    JOIN menu m
    ON s.product_id = m.product_id
    JOIN members mem
    ON s.customer_id = mem.customer_id
    WHERE s.order_date < mem.join_date
)
SELECT r.customer_id, r.product_name
FROM rank_items r
WHERE r.rank =1

 * sqlite://
Done.


[('A', 'sushi'), ('A', 'curry'), ('B', 'curry')]

**Answer-07**:

Last item purchased before becoming a member for:
- Customer A was curry and sushi (iteams were purchased in a same day)
- Customer B was sushi


### 8. What is the total items and amount spent for each member before they became a member?

In [20]:
%%sql
SELECT s.customer_id, COUNT(m.product_name) AS count_items, SUM(m.price) AS total_price
FROM sales s
JOIN menu m
ON s.product_id = m.product_id
JOIN members mem
ON s.customer_id = mem.customer_id
WHERE s.order_date < mem.join_date
GROUP BY s.customer_id

 * sqlite://
Done.


[('A', 2, 25), ('B', 3, 40)]

### 9. If each $1 spent equates to 10 points and sushi has a 2x points multiplier - how many points would each customer have?

In [21]:
%%sql
SELECT s.customer_id, 
SUM(CASE 
        WHEN m.product_name <> 'sushi' 
        THEN (m.price*10) 
        ELSE (m.price*20) 
    END ) AS total_points
FROM sales s 
JOIN menu m 
ON s.product_id = m.product_id
GROUP BY s.customer_id

 * sqlite://
Done.


[('A', 860), ('B', 940), ('C', 360)]

In [22]:
%%sql
SELECT s.customer_id, SUM(IIF(s.product_id = 1, m.price*20, m.price*10)) as total_points
FROM sales s
JOIN menu  m
ON s.product_id=m.product_id
GROUP BY s.customer_id

 * sqlite://
Done.


[('A', 860), ('B', 940), ('C', 360)]

<u>Answer-09</u>:

- Customer A has 860 points
- Customer B has 940 points
- Customer C has 360 points

### 10. In the first week after a customer joins the program (including their join date) they earn 2x points on all items, not just sushi - how many points do customer A and B have at the end of January?

In [27]:
%%sql
select sales.customer_id, 
SUM(IIF(sales.order_date >= DATEADD(DAY,7,members.join_date) and sales.order_date <= DATEADD(DAY,7,members.join_date), menu.price*20, menu.price*10)) as TotalPoints
from sales
inner join menu on sales.product_id=menu.product_id
inner join members on sales.customer_id=members.customer_id
group by sales.customer_id

 * sqlite://
(sqlite3.OperationalError) no such column: DAY
[SQL: select sales.customer_id, SUM(IIF(sales.order_date >= DATEADD(DAY,7,members.join_date) and sales.order_date <= DATEADD(DAY,7,members.join_date), menu.price*20, menu.price*10)) as TotalPoints
from sales
inner join menu on sales.product_id=menu.product_id
inner join members on sales.customer_id=members.customer_id
group by sales.customer_id]
(Background on this error at: http://sqlalche.me/e/14/e3q8)


In [1]:
# %%sql
# WITH sales_member_data AS 
#     (
#         SELECT s.customer_id,s.order_date, m.product_name,m.price, mem.join_date 
#         FROM sales s 
#         INNER JOIN members mem 
#         ON s.customer_id = mem.customer_id
#         INNER JOIN menu m
#         ON s.product_id= m.product_id
#     ),
# insights AS
#     (
#         SELECT customer_id,order_date,product_name,price,join_date,
#         CASE 
#         WHEN join_date > order_date  OR join_date IS null  THEN 'N'
#         ELSE 'Y'
#         END AS Member
#         FROM sales_member_data
#     ),
# order_before_becoming_members AS
#     (
#         SELECT * FROM insights 
#         WHERE Member = 'Y'
#     ),
# points_table AS
#     (
#         SELECT *,
#         CASE 
#         WHEN order_date between join_date AND DATEADD(DAY,7,join_date) THEN 20*price
#         WHEN order_date > DATEADD(DAY,7,join_date) AND product_name = 'sushi' THEN 20*price
#         ELSE 10*price
#         END AS points
#         FROM order_before_becoming_members
#     )
# SELECT customer_id,SUM(points) AS total_points 
# FROM points_table 
# WHERE MONTH(order_date) = 01 
# GROUP BY customer_id